src/details/ArborX_PredicateHelpers.hpp

```diff
@@ -0,0 +1,169 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_PREDICATE_HELPERS_HPP
+#define ARBORX_PREDICATE_HELPERS_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_GeometryTraits.hpp>
+#include <ArborX_HyperSphere.hpp>
+#include <ArborX_Predicates.hpp>
+
+namespace ArborX
+{
+namespace Experimental
+{
+
+template <typename UserPrimitives>
+struct PrimitivesIntersect
+{
+private:
+  using Primitives = Details::AccessValues<UserPrimitives, PrimitivesTag>;
+  // FIXME:
+  // using Geometry = typename Primitives::value_type;
+  // static_assert(GeometryTraits::is_valid_geometry<Geometry>{});
+
+public:
+  Primitives _primitives;
+};
+
+template <typename UserPrimitives>
+struct PrimitivesWithRadius
+{
+private:
+  using Primitives = Details::AccessValues<UserPrimitives, PrimitivesTag>;
+  using Point = typename Primitives::value_type;
+  static_assert(GeometryTraits::is_point<Point>::value);
+  using Coordinate = typename GeometryTraits::coordinate_type<Point>::type;
+
+public:
+  Primitives _primitives;
+  Coordinate _r;
+
+  PrimitivesWithRadius(UserPrimitives const &user_primitives, Coordinate r)
+      : _primitives(user_primitives)
+      , _r(r)
+  {}
+};
+
+template <class UserPrimitives>
+struct PrimitivesNearestK
+{
+private:
+  using Primitives = Details::AccessValues<UserPrimitives, PrimitivesTag>;
+  // FIXME:
+  // using Geometry = typename Primitives::value_type;
+  // static_assert(GeometryTraits::is_valid_geometry<Geometry>{});
+
+public:
+  Primitives _primitives;
+  int _k;
+};
+
+template <typename Primitives>
+auto intersect_geometries(Primitives const &primitives)
+{
+  Details::check_valid_access_traits(PrimitivesTag{}, primitives,
+                                     Details::DoNotCheckGetReturnType());
+  return PrimitivesIntersect<Primitives>{primitives};
+}
+
+template <typename Primitives, typename Coordinate>
+auto intersect_geometries_with_radius(Primitives const &primitives,
+                                      Coordinate r)
+{
+  Details::check_valid_access_traits(PrimitivesTag{}, primitives);
+  return PrimitivesWithRadius<Primitives>(primitives, r);
+}
+
+template <typename Primitives>
+auto nearest_k(Primitives const &primitives, int k)
```

Comment:
I'm open to suggestion how to name these. We could try to unify them with existing `ArborX::intersects` and `ArborX::nearest`.

Reply:
Need to think about it more.

Reply:
What about `make_nearest` and `make_intersects`?

Reply:
Not sure if `make_nearest` or `make_nearest_queries` is better.